agent/local/state.go

```diff
@@ -1362,9 +1372,23 @@ func (l *State) pruneCheck(id structs.CheckID) {
 	delete(l.checks, id)
 }
 
+// RegistrationTokenFallback returns a fallback function to be used when
+// determining the token to use for service sync.
+//
+// The fallback function will return the config file registration token if the
+// given service was sourced from a service definition in a config file.
+func (l *State) RegistrationTokenFallback(key structs.ServiceID) func() string {
```

Comment:
nit: I don't think this is used outside `package state` so it could be private.

Alternatively, what do you think about merging this logic into the body of `aclTokenForServiceSync` since it already does a lookup of `l.services[key]`?

Passing around a lock-guarded map in a closure makes me a little cautious; in this PR the codepaths are synchronized but this could be accidentally misused in the future.

Reply:
> I don't think this is used outside package state so it could be private.

I agree. I had to export the method because the test package is different (`local` vs `local_test`). Which seems unusual to me. But, without exporting it I can't call it in unit tests.

> Alternatively, what do you think about merging this logic into the body of aclTokenForServiceSync since it already does a lookup of l.services[key]?

I opted against this because `aclTokenForServiceSync` is also used in [deleteService](https://github.com/hashicorp/consul/blob/main/agent/local/state.go#L1269-L1274).

So, do we want deleteService to also incorporate the config file registration token in its list of fallback tokens? Generally, it seems better to me if it does not.

The main concern to me is if the config_file_registration token has been deleted, then it would fail to deregister the service and we'd see errors in logs. Also, it should fallback to using the agent token anyway. (It is able to use the agent token for service deregistrations because the Catalog.Deregister RPC accepts a token with the relevant node:write permissions).

Because the agent token must have node:write permissions (or else it could not have registered it's node into the catalog) and because the agent token is probably less likely to have been deleted (because agent lifecycle is longer than service instance lifecycle), it seems like the agent could skip straight to using the agent token without considering the config_file_registration token for the deregistration, rather than incorporating the agent token in the list of "fallback" tokens. And that would be faster and would not generate a deceptive log message. There's some relevant discussion on this here: https://github.com/hashicorp/consul/issues/8078

That's why I opted against inlining the config_file_registration token fallback into aclTokenForServiceSync. Although it does leave me a question: why does deleteService try using the service token and then fallback to agent token, instead of unilaterally using the agent token for deregistrations?

Reply:
My understanding is that the `agent` token typically only has `node:write` on itself. The agent token generally wouldn't have `service:write` needed to deregister a service.

Reply:
> My understanding is that the agent token typically only has node:write on itself. The agent token generally wouldn't have service:write needed to deregister a service.

Right, but the Consul servers will accept a deregistration if token contains node:write for the node containing that service. See https://github.com/hashicorp/consul/pull/5217 and https://github.com/hashicorp/consul/blob/275a0b8e7fa73296b3c1b27742fe4a44f0e62b13/agent/consul/catalog_endpoint.go#L441-L448

Reply:
Ah, interesting!

Reply:
> Although it does leave me a question: why does deleteService try using the service token and then fallback to agent token, instead of unilaterally using the agent token for deregistrations?

I've tested this.

The agent only falls back to the agent token if the service token is _unset_ for that particular service (i.e. `token` field is empty or absent from the service definition in any config files). It does not fallback to other tokens on failure to deregister (i.e. it will not try the deregistration with the service token and, on failure, then try the agent token).

If the original service token has been deleted from the servers, because the agent has stored that service token in its local state, it continues to use that original service token to deregister that service during each state sync - which will repeatedly fail each time the state sync is retried. This feels like a bit of a gotcha.

Basically, I'm weighing two options:

1. The existing behavior is "good", so we should have deleteService include the config_file_registration token in its list of fallbacks. If set, the config_file_registration would be used instead of the agent token. And if the config_file_registration token was deleted, then the deregistration would fail forever.
2. Or, the existing behavior isn't great, and it would be better for it to only use the agent token for service deregistrations because that will "just work" because of the node:write "bypass" for service deregistrations.

Thoughts @jkirschner-hashicorp?

Reply:
Is it accurate to say that the agent basically isn't functional if it lacks a token with `node:write` on itself?

If so... Is there any downside to approach 2? If I understand correctly, approach 2 would always work assuming the node is still functional (has a token the `node:write`).

Is there a separate case where a service is being deregistered directly from the server agents rather than from the node that owns the service, in cases where the node no longer exists but the service was never deregistered (but needs to be cleaned up)?

Happy to have a quick Zoom about this tomorrow.